File: src/statistics/loss.ts
/** Explicit frequency/severity assumptions and compound loss simulation. */

import { type SeverityAssumption, ValidationError } from "../models.js";
import { type Estimate, type Rows, estimateMetric, scenarioGroups, validateRows } from "./core.js";
import { mean, quantile } from "./dist.js";
import { Rng } from "./random.js";

export interface HarmLoss {
  status: "simulated" | "severity_assumption_required"; currency?: string; occurrenceFrequency?: Estimate; severityAssumption?: Record<string, unknown>;
  mean?: number; median?: number; p95?: number; p99?: number; tailMeanAboveP99?: number; expectedLossInterval?: [number, number]; predictiveInterval?: [number, number];
  frequencyModel?: Record<string, number | string>; intervalKind?: string;
}

export interface LossResult { functionId: string; horizonScenarios: number; simulations: number; seed: number; harms: Record<string, HarmLoss>; assumptions: string[] }

export function lossDistribution(rows: Rows, options: {
  functionId: string; severityAssumptions: Record<string, SeverityAssumption>; simulations?: number; seed?: number; horizon?: number;
}): LossResult {
  const { functionId, severityAssumptions, simulations = 5000, seed = 0, horizon = 10000 } = options;
  const selected = rows.filter((r) => r.functionId === functionId);
  validateRows(selected);
  if (simulations < 100 || horizon < 1) throw new ValidationError("Use at least 100 simulations and a positive horizon");
  const rng = new Rng(seed);
  const groups = scenarioGroups(selected);
  const clusterSizes = new Map<string, number>();
  for (const trials of groups.values()) clusterSizes.set(trials[0].cluster, (clusterSizes.get(trials[0].cluster) ?? 0) + 1);
  const weights = [...clusterSizes.values()];
  const total = weights.reduce((a, b) => a + b, 0);
  const effectiveClusters = total ** 2 / weights.reduce((a, b) => a + b * b, 0);
  const harms = [...new Set([...selected.flatMap((r) => Object.keys(r.occurred ?? {})), ...Object.keys(severityAssumptions)])].sort();
  const output: Record<string, HarmLoss> = {};
  for (const harm of harms) {
    const assumption = severityAssumptions[harm];
    if (!assumption) { output[harm] = { status: "severity_assumption_required" }; continue; }
    const kind = assumption.distribution ?? "fixed";
    const meanSeverity = Number(assumption.mean ?? assumption.amount ?? 0);
    if (!Number.isFinite(meanSeverity) || meanSeverity < 0) throw new ValidationError("Assumed severity mean must be finite and nonnegative");
    if (!assumption.currency) throw new ValidationError("Every severity assumption must state its currency");
    if (!["fixed", "gamma", "lognormal"].includes(kind)) throw new ValidationError("Severity distribution must be fixed, gamma or lognormal");
    const cv = Number(assumption.coefficientOfVariation ?? 1);
    if (!Number.isFinite(cv) || cv <= 0) throw new ValidationError("coefficientOfVariation must be positive and finite");
    const frequency = mean([...groups.values()].map((trials) => mean(trials.map((r) => (r.occurred?.[harm] ? 1 : 0)))));
    // Clusters, not repeated trials, carry the frequency information: an explicitly assumed conservative beta model.
    const a = 0.5 + frequency * effectiveClusters, b = 0.5 + (1 - frequency) * effectiveClusters;
    const totals = new Float64Array(simulations), expected = new Float64Array(simulations);
    for (let s = 0; s < simulations; s++) {
      const f = rng.beta(a, b);
      const count = rng.binomial(horizon, f);
      let loss = count * meanSeverity;
      if (meanSeverity > 0 && kind === "gamma" && count > 0) {
        const shape = 1 / cv ** 2;
        loss = rng.gamma(count * shape, meanSeverity / shape);
      } else if (meanSeverity > 0 && kind === "lognormal") {
        const sigma = Math.sqrt(Math.log1p(cv ** 2)), mu = Math.log(meanSeverity) - sigma ** 2 / 2;
        loss = 0;
        for (let k = 0; k < count; k++) loss += rng.lognormal(mu, sigma);
      }
      totals[s] = loss;
      expected[s] = f * horizon * meanSeverity;
    }
    const p99 = quantile(totals, 0.99);
    output[harm] = {
      status: "simulated", currency: assumption.currency, occurrenceFrequency: estimateMetric(selected, `occurrences:${harm}`, { seed }),
      severityAssumption: { ...assumption, distribution: kind, mean: meanSeverity },
      mean: mean(totals), median: quantile(totals, 0.5), p95: quantile(totals, 0.95), p99,
      tailMeanAboveP99: mean(Array.from(totals).filter((v) => v >= p99)),
      expectedLossInterval: [quantile(expected, 0.025), quantile(expected, 0.975)],
      predictiveInterval: [quantile(totals, 0.025), quantile(totals, 0.975)],
      frequencyModel: { family: "beta", priorAlpha: 0.5, priorBeta: 0.5, posteriorAlpha: a, posteriorBeta: b, effectiveIndependentClusters: effectiveClusters },
      intervalKind: "95% assumption-dependent posterior credible/predictive intervals",
    };
  }
  return { functionId, horizonScenarios: horizon, simulations, seed, harms: output,
    assumptions: ["Severity distributions are assumptions, not inferred guarantees.",
      "Frequency uses an assumed beta model with effective top-level cluster counts and Jeffreys prior.",
      "Future occurrence counts are conditionally binomial; clustered bursts beyond parameter uncertainty are not modeled.",
      "Severity is independent of frequency and scenarios within each harm class.",
      "Harm classes are reported separately because the same occurrence may satisfy multiple graders.",
      "Zero observed events retain nonzero frequency uncertainty."] };
}

File: src/signal_sdk/statistics/loss.py
"""Explicit frequency/severity assumptions and compound loss simulation."""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any

import numpy as np

from .core import Rows, trajectory_groups, estimate_metric, validate_rows


def loss_distribution(rows: Rows, *, function_id: str, severity_assumptions: Mapping[str, Mapping[str, Any]],
                      simulations: int = 5000, seed: int = 0, horizon: int = 10000) -> dict[str, Any]:
    selected = [r for r in rows if r["function_id"] == function_id]
    validate_rows(selected)
    if simulations < 100 or horizon < 1:
        raise ValueError("Use at least 100 simulations and a positive horizon")
    rng = np.random.default_rng(seed)
    groups = trajectory_groups(selected)
    cluster_sizes: dict[str, int] = {}
    for trials in groups.values():
        cluster = str(trials[0]["cluster"])
        cluster_sizes[cluster] = cluster_sizes.get(cluster, 0) + 1
    weights = np.array(list(cluster_sizes.values()), dtype=float)
    effective_clusters = float(weights.sum() ** 2 / np.sum(weights ** 2))
    harms = sorted({h for r in selected for h in r.get("occurred", {})} | set(severity_assumptions))
    output = {}
    for harm in harms:
        if harm not in severity_assumptions:
            output[harm] = {"status": "severity_assumption_required"}
            continue
        assumptions = dict(severity_assumptions[harm])
        kind = assumptions.get("distribution", "fixed")
        mean = float(assumptions.get("mean", assumptions.get("amount", 0)))
        if not np.isfinite(mean) or mean < 0:
            raise ValueError("Assumed severity mean must be finite and nonnegative")
        if not assumptions.get("currency"):
            raise ValueError("Every severity assumption must state its currency")
        if kind not in {"fixed", "gamma", "lognormal"}:
            raise ValueError("Severity distribution must be fixed, gamma or lognormal")
        cv = float(assumptions.get("coefficient_of_variation", 1))
        if not np.isfinite(cv) or cv <= 0:
            raise ValueError("coefficient_of_variation must be positive and finite")
        trajectory_rates = [np.mean([bool(r.get("occurred", {}).get(harm, False)) for r in trials]) for trials in groups.values()]
        frequency = float(np.mean(trajectory_rates))
        # Clusters, not repeated trials, determine frequency information.
        # Fractional pseudo-counts form an explicitly assumed conservative
        # beta model; these are posterior predictions, not bootstrap CIs.
        a, b = .5 + frequency * effective_clusters, .5 + (1 - frequency) * effective_clusters
        frequencies = rng.beta(a, b, simulations)
        counts = rng.binomial(horizon, frequencies)
        totals = counts.astype(float) * mean
        if mean > 0 and kind == "gamma":
            shape = 1 / cv ** 2
            active = counts > 0
            totals[active] = rng.gamma(counts[active] * shape, mean / shape)
        elif mean > 0 and kind == "lognormal":
            sigma = np.sqrt(np.log1p(cv ** 2))
            mu = np.log(mean) - sigma ** 2 / 2
            for index, count in enumerate(counts):
                totals[index] = float(rng.lognormal(mu, sigma, int(count)).sum())
        expected = frequencies * horizon * mean
        output[harm] = {"status": "simulated", "currency": assumptions["currency"],
                        "occurrence_frequency": estimate_metric(selected, f"occurrences:{harm}", seed=seed),
                        "severity_assumption": {**assumptions, "distribution": kind, "mean": mean},
                        "mean": float(np.mean(totals)), "median": float(np.median(totals)),
                        "p95": float(np.quantile(totals, .95)), "p99": float(np.quantile(totals, .99)),
                        "tail_mean_above_p99": float(np.mean(totals[totals >= np.quantile(totals, .99)])),
                        "expected_loss_interval": list(map(float, np.quantile(expected, [.025, .975]))),
                        "predictive_interval": list(map(float, np.quantile(totals, [.025, .975]))),
                        "frequency_model": {"family": "beta", "prior_alpha": .5, "prior_beta": .5,
                                            "posterior_alpha": a, "posterior_beta": b,
                                            "effective_independent_clusters": effective_clusters},
                        "interval_kind": "95% assumption-dependent posterior credible/predictive intervals"}
    return {"function_id": function_id, "horizon_trajectories": horizon, "simulations": simulations, "seed": seed,
            "harms": output,
            "assumptions": ["Severity distributions are assumptions, not inferred guarantees.",
                            "Frequency uses an assumed beta model with effective top-level cluster counts and Jeffreys prior.",
                            "Future occurrence counts are conditionally binomial; clustered bursts beyond parameter uncertainty are not modeled.",
                            "Severity is independent of frequency and trajectories within each harm class.",
                            "Harm classes are reported separately because the same occurrence may satisfy multiple graders.",
                            "Zero observed events retain nonzero frequency uncertainty."]}
